File: aws/lambda_function/get_kindle_recommendations.py
import json
import os
from pymongo import MongoClient

def lambda_handler(event, context):
    # 1. ESTRAZIONE DATI DALLA RICHIESTA DELL'APP
    if isinstance(event.get('body'), str):
        body = json.loads(event['body'])
    else:
        body = event.get('body', event)
        
    kindle_genres = body.get('kindle_genres', [])
    watched_video_ids = body.get('watched_video_ids', []) # Bug fixato: ora lo prende dal payload in modo sicuro
    
    if not kindle_genres:
        return {
            'statusCode': 400,
            'headers': {'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'},
            'body': json.dumps({"error": "Parametro 'kindle_genres' mancante o vuoto."})
        }

    # 2. DIZIONARIO DI MAPPATURA (KINDLE -> TEDX)
    mapping_bridge = {
        "Scienza e Tecnologia": ["science", "technology", "innovation", "future"],
        "Biografie": ["biography", "history", "personal growth", "identity"],
        "Saggistica": ["society", "culture", "politics", "global issues"],
        "Self-help": ["psychology", "mental health", "happiness", "motivation"],
        "Economia": ["business", "economics", "finance", "work"],
        "Thriller": ["crime", "mystery", "human nature"]
    }
    
    tedx_tags_to_search = []
    for genre in kindle_genres:
        if genre in mapping_bridge:
            tedx_tags_to_search.extend(mapping_bridge[genre])
            
    unique_tags = list(set(tedx_tags_to_search))
    
    # Se non c'è nessun match con i tag, ritorno lista vuota
    if not unique_tags:
        return {
            'statusCode': 200,
            'headers': {'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'},
            'body': json.dumps({"recommendations": [], "message": "Nessun tag corrispondente trovato."})
        }

    # 3. CONNESSIONE A MONGODB
    mongo_uri = os.environ.get('MONGO_URI')
    if not mongo_uri:
        return {
            'statusCode': 500,
            'body': json.dumps({"error": "MONGO_URI non configurata nelle variabili d'ambiente."})
        }
        
    client = MongoClient(mongo_uri)
    db = client['unibg_tedx_2026']
    talks_col = db['tedx_data']
    
    # 4. INTERROGAZIONE DEL DATA WAREHOUSE
    # Cerca i video che hanno i tag richiesti, escludendo quelli che l'utente ha già visto
    query = {
        "tags": {"$in": unique_tags}, 
        "_id": {"$nin": watched_video_ids}
    }
    
    # Prende massimo 5 risultati per mantenere l'app veloce
    results = talks_col.find(query).limit(5)
    
    recommended_talks = []
    for talk in results:
        recommended_talks.append({
            "id": talk.get('_id'),
            "title": talk.get('title'),
            "slug": talk.get('slug'),
            "url": talk.get('url'),
            "description": talk.get('description', ''),
            "duration": talk.get('duration')
        })
        
    # 5. RISPOSTA AL FRONTEND FLUTTER
    return {
        'statusCode': 200,
        'headers': {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*' # Essenziale per evitare errori CORS
        },
        'body': json.dumps({
            'tags_used_for_search': unique_tags,
            'recommendations': recommended_talks
        })
    }

File: backend_aws/lambdas/watch_next/get_kindle_recommendations.mjs
import { MongoClient } from 'mongodb';

const uri = process.env.MONGO_URI;
let cachedClient = null;
let cachedDb = null;

async function connectToDatabase() {
    if (cachedClient && cachedDb) return { client: cachedClient, db: cachedDb };
    if (!uri) throw new Error("La variabile d'ambiente MONGO_URI non è configurata.");
    
    const client = new MongoClient(uri);
    await client.connect();
    const db = client.db('unibg_tedx_2026');
    
    cachedClient = client;
    cachedDb = db;
    return { client, db };
}

export const handler = async (event) => {
    try {
        // ESTRAZIONE DATI DIRETTAMENTE DALL'EVENTO (Mapping Template)
        const kindleGenres = event.kindle_genres || [];
        const watchedVideoIds = event.watched_video_ids || [];

        if (kindleGenres.length === 0) {
            return {
                statusCode: 400,
                headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' },
                body: JSON.stringify({ error: "Parametro 'kindle_genres' mancante o vuoto nell'evento." })
            };
        }

        // DIZIONARIO DI MAPPATURA (KINDLE -> TEDX)
        const mappingBridge = {
            "Scienza e Tecnologia": ["science", "technology", "innovation", "future"],
            "Biografie": ["biography", "history", "personal growth", "identity"],
            "Saggistica": ["society", "culture", "politics", "global issues"],
            "Self-help": ["psychology", "mental health", "happiness", "motivation"],
            "Economia": ["business", "economics", "finance", "work"],
            "Thriller": ["crime", "mystery", "human nature"]
        };

        let tedxTagsToSearch = [];
        kindleGenres.forEach(genre => {
            if (mappingBridge[genre]) {
                tedxTagsToSearch.push(...mappingBridge[genre]);
            }
        });

        const uniqueTags = [...new Set(tedxTagsToSearch)];

        if (uniqueTags.length === 0) {
            return {
                statusCode: 200,
                headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' },
                body: JSON.stringify({ recommendations: [], message: "Nessun tag corrispondente trovato." })
            };
        }

        // CONNESSIONE E QUERY SU MONGODB ATLAS
        const { db } = await connectToDatabase();
        const talksCol = db.collection('tedx_data');

        const stringWatchedIds = watchedVideoIds.map(id => id.toString());

        const query = {
            tags: { $in: uniqueTags },
            _id: { $nin: stringWatchedIds }
        };

        const projection = { _id: 1, title: 1, slug: 1, url: 1, description: 1, duration: 1 };
        
        const cursor = talksCol.find(query).project(projection).limit(5);
        const recommendedTalks = [];

        await cursor.forEach((doc) => {
            recommendedTalks.push({
                id: doc._id,
                title: doc.title,
                slug: doc.slug,
                url: doc.url,
                description: doc.description || '',
                duration: doc.duration
            });
        });

        // RISPOSTA AL FRONTEND FLUTTER
        return {
            statusCode: 200,
            headers: {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            body: JSON.stringify({
                tags_used_for_search: uniqueTags,
                recommendations: recommendedTalks
            })
        };

    } catch (error) {
        console.error("Errore nella Lambda get_kindle_recommendations:", error);
        return {
            statusCode: 500,
            headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' },
            body: JSON.stringify({ error: "Errore interno del server", dettagli: error.message })
        };
    }
};